wallet/src/wallet/mod.rs

```diff
@@ -2377,6 +2402,104 @@ impl Wallet {
+    }
+
+    /// List unspent outpoints that are currently locked.
+    pub fn list_locked_unspent(&self) -> impl Iterator<Item = OutPoint> + '_ {
+        self.list_unspent()
+            .filter(|output| self.is_outpoint_locked(output.outpoint))
+            .map(|output| output.outpoint)
+    }
+
+    /// Whether the `outpoint` is currently locked. See [`Wallet::lock_outpoint`] for more.
+    pub fn is_outpoint_locked(&self, outpoint: OutPoint) -> bool {
+        if let Some(utxo_lock) = self.locked_outpoints.get(&outpoint) {
+            if utxo_lock.is_locked {
+                return utxo_lock
+                    .expiration_height
+                    .map_or(true, |height| self.chain.tip().height() < height);
+            }
+        }
+        false
+    }
+
+    /// Lock a wallet output identified by the given `outpoint`.
+    ///
+    /// A locked UTXO will not be selected as an input to fund a transaction. This is useful
+    /// for excluding or reserving candidate inputs during transaction creation. You can optionally
+    /// specify the `expiration_height` of the lock that defines the height of the local chain at
+    /// which the outpoint becomes spendable.
+    ///
+    /// You must persist the staged change for the lock status to be persistent. To unlock a
+    /// previously locked outpoint, see [`Wallet::unlock_outpoint`].
+    pub fn lock_outpoint(&mut self, outpoint: OutPoint, expiration_height: Option<u32>) {
+        use alloc::collections::btree_map;
+        let lock_value = true;
+        let mut changeset = locked_outpoints::ChangeSet::default();
+
+        // If the lock status changed, update the entry and record the change
+        // in the changeset.
+        match self.locked_outpoints.entry(outpoint) {
+            btree_map::Entry::Occupied(mut e) => {
+                let utxo = e.get_mut();
+                if !utxo.is_locked {
+                    utxo.is_locked = lock_value;
+                    changeset.locked_outpoints.insert(outpoint, lock_value);
+                }
+                if utxo.expiration_height != expiration_height {
+                    utxo.expiration_height = expiration_height;
+                    changeset
+                        .expiration_heights
+                        .insert(outpoint, expiration_height);
+                }
+            }
+            btree_map::Entry::Vacant(e) => {
+                e.insert(UtxoLock {
+                    outpoint,
+                    is_locked: lock_value,
+                    expiration_height,
+                });
+                changeset.locked_outpoints.insert(outpoint, lock_value);
+                changeset
+                    .expiration_heights
+                    .insert(outpoint, expiration_height);
+            }
+        };
+
+        self.stage.merge(ChangeSet {
+            locked_outpoints: changeset,
+            ..Default::default()
+        });
+    }
+
+    /// Unlock the wallet output of the specified `outpoint`.
```

Comment:
You should also warn the users as you did with `lock_outpoint` that these changes need to be persisted.